operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,377 @@
+	if regularExists {
+		// Check if the deleted object was for a regular DNS name. If so, add the DNS name
+		// to dnsNameList slice. Additionally, remove the corresponding entry of the object
+		// from regularObjInfo map.
+		matchesRegular = true
+		dnsNameList = append(dnsNameList, regularDNSName)
+		delete(res.regularObjInfo, objName)
+	} else {
+		// If the deleted object was not for a regular DNS name, then check if the deleted
+		// object was for a wildcard DNS name. If so, add the DNS names matching the wildcard
+		// DNS name to dnsNameList slice. Additionally, remove the corresponding entry of the
+		// object from wildcardObjInfo map.
+		wildcardDNSNames, wildcardExists := res.wildcardObjInfo[objName]
+		if !wildcardExists {
+			return
+		}
+		dnsNameList = append(dnsNameList, wildcardDNSNames.UnsortedList()...)
+		delete(res.wildcardObjInfo, objName)
+		matchesRegular = false
+	}
+
+	// Iterate through the dnsNameList slice.
+	for _, dnsName := range dnsNameList {
+		// Get the resolved name details corresponding to the DNS name.
+		resolvedName, exists := res.dnsNames[dnsName]
+		// If the corresponding resolved name details is not found,
+		// continue with the next DNS name.
+		if !exists {
+			continue
+		}
+
+		// If the object was for a regular DNS name, then unset the regularObjName
+		// field of the resolved name. Otherwise, unset the wildcardObjName field
+		// of the resolved name.
+		if matchesRegular {
+			resolvedName.regularObjExists = false
+		} else {
+			resolvedName.wildcardObjExists = false
+		}
+
+		// If both the regularObjName and wildcardObjName fields are unset, then
+		// there are no DNSNameResolver object that references the DNS name.
+		// Remove the details of the resolved name corresponding to the DNS name
+		// from the dnsNames map. Also, send the DNS name to the deleted channel
+		// indicating that the details corresponding to the DNS name have been
+		// deleted. Otherwise, add the updated resolved name to the dnsNames map
+		// corresponding to the DNS name.
+		if !resolvedName.regularObjExists && !resolvedName.wildcardObjExists {
+			delete(res.dnsNames, dnsName)
+			res.deleted <- dnsName
+		} else {
+			res.dnsNames[dnsName] = resolvedName
+		}
+	}
+}
+
+// getNextDNSNameDetails returns the DNS name with minimum next lookup time.
+// It also returns the next lookup time and the number of IP addresses
+// associated with the DNS name. If no such DNS name exists, then false is
+// returned, otherwise true is returned.
+func (res *Resolver) getNextDNSNameDetails() (string, time.Time, int, bool) {
+	res.dnsLock.Lock()
+	defer res.dnsLock.Unlock()
+
+	exists := false
+	var minNextLookupTime time.Time
+	var dns string
+	var numIPs int
+
+	for dnsName, resolvedName := range res.dnsNames {
+		if !exists || resolvedName.minNextLookupTime.Before(minNextLookupTime) {
+			exists = true
+			minNextLookupTime = resolvedName.minNextLookupTime
+			dns = dnsName
+			numIPs = resolvedName.numIPs
```

Comment:
If you return here, then you might not actually get the `dnsName` that has the lowest `minNextLookupTime`.  Adding a `return` here causes `TestResolver` to fail.

Reply:
This function is used to get the DNS name with the `minNextLookupTime` so that we can look it up in the next round. Every round we need to do this exercise to find out the DNS name with `minNextLookupTime` and then send DNS lookup request for the same.